lib/kube/proxy/pod_rbac_test.go

```diff
@@ -0,0 +1,539 @@
+	type args struct {
+		user      types.User
+		namespace string
+	}
+	type want struct {
+		listPodsResult   []string
+		getTestPodResult error
+	}
+	tests := []struct {
+		name string
+		args args
+		want want
+	}{
+		{
+			name: "list default namespace pods for user with full access",
+			args: args{
+				user:      userWithFullAccess,
+				namespace: metav1.NamespaceDefault,
+			},
+			want: want{
+				listPodsResult: []string{
+					"default/nginx-1",
+					"default/nginx-2",
+					"default/test",
+				},
+			},
+		},
+		{
+			name: "list pods in every namespace for user with full access",
+			args: args{
+				user:      userWithFullAccess,
+				namespace: metav1.NamespaceAll,
+			},
+			want: want{
+				listPodsResult: []string{
+					"default/nginx-1",
+					"default/nginx-2",
+					"default/test",
+					"dev/nginx-1",
+					"dev/nginx-2",
+				},
+			},
+		},
+		{
+			name: "list default namespace pods for user with default namespace",
+			args: args{
+				user:      userWithNamespaceAccess,
+				namespace: metav1.NamespaceDefault,
+			},
+			want: want{
+				listPodsResult: []string{
+					"default/nginx-1",
+					"default/nginx-2",
+					"default/test",
+				},
+			},
+		},
+		{
+			name: "list pods in every namespace for user with default namespace",
+			args: args{
+				user:      userWithNamespaceAccess,
+				namespace: metav1.NamespaceAll,
+			},
+			want: want{
+				listPodsResult: []string{
+					"default/nginx-1",
+					"default/nginx-2",
+					"default/test",
+				},
+			},
+		},
+		{
+			name: "list default namespace pods for user with limited access",
+			args: args{
+				user:      userWithLimitedAccess,
+				namespace: metav1.NamespaceDefault,
+			},
+			want: want{
+				listPodsResult: []string{
+					"default/nginx-1",
+					"default/nginx-2",
+				},
+				getTestPodResult: &kubeerrors.StatusError{
+					ErrStatus: metav1.Status{
+						Status:  "Failure",
+						Message: "[00] access denied",
+						Code:    403,
+					},
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			// generate a kube client with user certs for auth
+			client, _ := testCtx.genTestKubeClientTLSCert(
+				t,
+				tt.args.user.GetName(),
+				kubeCluster,
+			)
+
+			rsp, err := client.CoreV1().Pods(tt.args.namespace).List(
+				testCtx.ctx,
+				metav1.ListOptions{},
+			)
+			require.NoError(t, err)
+
+			require.Equal(t, tt.want.listPodsResult, getPodsFromPodList(rsp.Items))
+
+			_, err = client.CoreV1().Pods(metav1.NamespaceDefault).Get(
+				testCtx.ctx,
+				testPodName,
+				metav1.GetOptions{},
+			)
+			require.Equal(t, tt.want.getTestPodResult, err)
+		})
+	}
+}
+
+func getPodsFromPodList(items []corev1.Pod) []string {
+	pods := make([]string, 0, len(items))
+	for _, item := range items {
+		pods = append(pods, filepath.Join(item.Namespace, item.Name))
+	}
+	return pods
+}
+
+func TestWatcherResponseWriter(t *testing.T) {
+	t.Parallel()
+	fakeEvents := fakeEvents()
```

Comment:
Suggestion: I would rather inline fakeEvents here than use the method. We take slices from it, so the contents are strongly tied to the rest of the test. If it's reused, I'd rather duplicate the declarations as needed, so tests are clearer and have room to evolve separately.